dbms/src/Storages/KVStore/Utils/AsyncTasks.h

```diff
@@ -14,110 +14,358 @@
             return true;
         }
         return false;
     }
 
-    bool addTask(Key k, Func f)
+    // Safety: Throws if
+    // 1. The task not exist and `throw_if_noexist`.
+    // 2. Throw in `result_dropper`.
+    /// Usage:
+    /// 1. If the task is in `Finished` state
+    ///     It's result will be cleaned with `result_dropper`.
+    /// 2. If the task is in `Running` state
+    ///     Make sure the executor will do the clean.
+    /// 3. If the tasks is in `InQueue` state
+    ///     The task will directly return when it's eventually run by a thread.
+    /// 4. If the tasks is in `NotScheduled` state
+    ///     `throw_if_noexist` controls whether to throw.
+    /// NOTE: The task element will be removed after calling this function.
+    template <typename ResultDropper>
+    TaskState asyncCancelTask(Key k, ResultDropper result_dropper, bool throw_if_noexist)
+    {
+        auto cancel_handle = getCancelHandleFromCaller(k, throw_if_noexist);
+        if (cancel_handle)
+        {
+            cancel_handle->doCancel();
+            // Cancel logic should do clean itself
+        }
+
+        auto state = queryState(k);
+        if (!throw_if_noexist && state == TaskState::NotScheduled)
+            return state;
+        if (state == TaskState::Finished)
+        {
+            result_dropper();
+        }
+
+        // `result_dropper` may remove the task by `fetchResult`.
+        leakingDiscardTask(k);
+
+        return state;
+    }
+
+    TaskState asyncCancelTask(Key k)
     {
+        return asyncCancelTask(
+            k,
+            []() {},
+            true);
+    }
+
+    // Safety: Throws if
+    // 1. The task is not found, and throw_on_no_exist.
+    // 2. The cancel_handle is already set.
+    // 3. Throw in `result_dropper`.
+    /// Usage:
+    /// 1. If the task is in `Finished`/`Running` state
+    ///     It's result is returned. The Caller may do the rest cleaning.
+    /// 3. If the tasks is in `InQueue` state
+    ///     The task will directly return when it's eventually run by a thread.
+    /// 4. If the tasks is in `NotScheduled` state
+    ///     It will throw.
+    /// Returns:
+    /// 1. `NotScheduled` or `InQueue`
+    /// 2. `R`
+    /// 3. Exception
+    /// NOTE: The task element will be removed after calling this function.
+    [[nodiscard]] TaskState blockedCancelRunningTask(Key k, bool throw_on_no_exist = true)
+    {
+        auto cancel_handle = getCancelHandleFromCaller(k);
+        auto state = queryState(k);
+        if (state == TaskState::NotScheduled)
+        {
+            if (throw_on_no_exist)
+            {
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't block wait a non-scheduled task");
+            }
+            else
+            {
+                return state;
+            }
+        }
+
+        // Only one thread can block cancel and wait.
+        RUNTIME_CHECK_MSG(!cancel_handle->canceled(), "Try block cancel running task twice");
```

Comment:
why we need this limitation?

Reply:
Because the `AsyncTasks` only supports one consumer, so why would this only consumer cancel twice? I think it is a mistake.